tests/common/pkt_filter/filter_pkt_in_buffer.py

```diff
@@ -0,0 +1,232 @@
+
+    Args:
+        pkt: Scapy packet
+
+    Returns:
+        Packet dictionary
+    """
+    packet_dict = OrderedDict()
+    counter = 0
+
+    while True:
+        layer = pkt.getlayer(counter)
+        if not layer:
+            break
+
+        layer_dict = _parse_layer(layer)
+        if layer_dict:
+            packet_dict.update(layer_dict)
+
+        counter += 1
+
+    return packet_dict
+
+
+class FilterPktBuffer(object):
+    """
+    FilterPktBuffer class for finding of packets in the buffer of PTF
+    """
+    def __init__(self, ptfadapter, exp_pkt, dst_port_number, match_fields=None, ignore_fields=None):
+        """
+        Initialize an object for finding packets in the buffer
+
+        Args:
+            ptfadapter: PTF adapter
+            exp_pkt: Expected packet
+            dst_port_number: Destination port number
+            match_fields: List of packet fields that should be matched
+            ignore_fields: List of packet fields that should be ignored
+        """
+        self.received_pkt = None
+        self.received_pkt_diff = []
+        self.ptfadapter = ptfadapter
+        self.pkt = exp_pkt
+        self.dst_port_number = dst_port_number
+
+        if match_fields is None:
+            match_fields = []
+        self.match_fields = match_fields
+
+        if ignore_fields is None:
+            ignore_fields = []
+        self.ignore_fields = ignore_fields
+
+        self.masked_exp_pkt = mask.Mask(self.pkt)
+        self.pkt_dict = convert_pkt_to_dict(self.pkt)
+
+
+    def __ignore_fields(self):
+        """
+        Ignore fields of packet
+        """
+        for field, value in self.ignore_fields:
+            self.masked_exp_pkt.set_do_not_care_scapy(getattr(packet, field), value)
+
+
+    def __remove_ignore_fields(self, pkt_dict):
+        """
+        Remove ignored fields from packet dictionary
+
+        Args:
+            pkt_dict: Packet dictionary
+
+        Returns:
+            Packet dictionary without ignored fields
+        """
+        pkt_dict = copy.deepcopy(pkt_dict)
+
+        for field, value in self.ignore_fields:
+            if pkt_dict.get(field):
+                pkt_dict[field].pop(value)
+
+        return pkt_dict
+
+
+    def __find_pkt_in_buffer(self):
+        """
+        Find expected packet in buffer by using matched fields
+
+        Returns:
+            Received packet
+        """
+        time.sleep(0.1)
+        common_buffer = self.ptfadapter.dataplane.packet_queues
+        packet_buffer = common_buffer[(0, self.dst_port_number)][:]
+
+        for pkt in packet_buffer:
+            packet_dict = convert_pkt_to_dict(Ether(pkt[0]))
+
+            for field, value in self.match_fields:
+                try:
+                    if packet_dict[field][value] != self.pkt_dict[field][value]:
+                        break
+                except KeyError:
+                    break
+            else:
+                return Ether(pkt[0])
```

Comment:
Only the first packet matches the `exp_pkt` will be returned? What if there are multiple packets in buffer match the `exp_pkt`?

Reply:
The method returns the first packet which matches the exp_pkt. Do we need to return multiple packets?

Reply:
I think yes, then duplicated packets can be detected.